# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def addTwo(x):
    return x + 2

print(addTwo(7))

# Lambda function ex.
# lambda var : (return line)

print((lambda x : x + 2)(8)) # the () executes the function and acts as what you want to use


9
10


#### Saving to a Variable

In [2]:
func = lambda x : x + 2

func(12)

14

#### Multiple Inputs

In [3]:
print((lambda x, y, z : x * y * z)(3,5,8)) # Self invoking

# multiple inputs saved to a variable

func2 = lambda a,b,c : a*b*c
print(func2(3,5,8))

120
120


#### Passing a Lambda into a Function

In [4]:
def multiply(f, num):
    '''
    f expects a lambda function
    num expects a number
    '''
    
    return f(num)

print(multiply(lambda x : x*x, 7))

49


#### Returning a Lambda from a Function

In [5]:
def mult(num):
    return num * 4

#function within a function
def returnFunc():
    def innerFunc(num):
        return num * 15
    return innerFunc


refToInner = returnFunc()
print(refToInner(6))
        
    
    
def returnLamb(b,c):
    return lambda x,y : x + y + b + c

refToLamb = returnLamb(5,10)
print(refToLamb(7,9))

90
31


#### If Statements within Lambdas

In [6]:
# SYNTAX
# lambda x : True if (condition) else False

f = lambda num : num * 2 if num < 10 else num + 2
print(f(25))
print(f(6))
print(f(10))

# def doubleIfSmolElseAdd(num):
#     if num > 10:
#         return num * 2
#     else:
#         return num + 2
    


27
12
12


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [7]:
def triple():
    return lambda num : num ** 3

f = triple()
print(f(6))

216


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [8]:
# SYNTAX
# map(func, iter)

def square(num):
    if num < 10 and num % 2 == 0:
        return num ** 2
    else:
        return num

l1 = [1,2,3,4]
l2 = [5,54,5,6]

result1 = list(map(square, l1))
print(result1)


[1, 4, 3, 16]


#### Using Lambda's with Map

In [9]:
result2 = list(map(lambda num : num **2 if num < 10 and num % 2 == 0 else num,l2))
print(result2)

[5, 54, 5, 36]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [10]:
result3 = list(map(lambda num : (num * 2)-1, l2))
print(result3)

''.join(list(map(lambda letter: letter.upper() if letter.lower() in {'a', 'e', 'i', 'o', 'u'} else letter.lower(), 'DIANTe')))

[9, 107, 9, 11]


'dIAntE'

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [11]:
# SYNTAX:
# filter(func, iter)

def a_names(name):
    if name[0].lower() == 'a':
        return True
    else:
        return False

names = ['diante', 'brandt', 'amy', 'ashley', 'mary', 'bobby']

list(filter(a_names, names))


    

['amy', 'ashley']

#### Using Lambda's with Filter()

In [12]:
list(filter(lambda name : name[0].upper() == 'B', names))

['brandt', 'bobby']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [34]:
from statistics import mean

l1 = [17,45,66,21,45,33,54]

result = list(filter(lambda num : num if num > mean(l1) else False,l1)) # need to add else to complete function
print(result)


[45, 66, 45, 54]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [37]:
from functools import reduce
# reduce (func, iter)

l1 = [2,4,6,8,22]

def addNums(num1, num2):
    return num1 + num2

res = reduce(addNums, l1) #can only take in two NO MORE
print(res)

42


#### Using Lambda's with Reduce()

In [38]:
reduce(lambda x , y: x-y, l1)

-38

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [43]:
def addNums(num):
    # base case (answer for where you loop stops)
    if num <= 1:
        print('addNums(1) = 1')
        return num
    # recursive case (what you want to continue to happen and loop)
    else:
        print(f'addNums({num}) = {num} + addNums({num-1})')
        return num + addNums(num-1)
        

    
print(addNums(7))

addNums(7) = 7 + addNums(6)
addNums(6) = 6 + addNums(5)
addNums(5) = 5 + addNums(4)
addNums(4) = 4 + addNums(3)
addNums(3) = 3 + addNums(2)
addNums(2) = 2 + addNums(1)
addNums(1) = 1
28


#### Writing a Factorial Function

In [44]:
# 5! = 5 * 4 * 3 * 2 * 1
def factorial(num):
    if num <=1: 
        return 1
    else:
        return num * factorial(num-1)
    
print(factorial(6))

720


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [1]:
def myRange(start, stop, step):
    while start < stop:
        yield start # yield keyword denotes a generator
        start += step
        
for num in myRange(2,100,3):
    print(num)

2
5
8
11
14
17
20
23
26
29
32
35
38
41
44
47
50
53
56
59
62
65
68
71
74
77
80
83
86
89
92
95
98


#### Infinite Generator

In [14]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [51]:
def squareGen(num):
    while num > 0:
        yield num ** 2
        num -= 1
        
for num in squareGen(10):
    print(num)

100
81
64
49
36
25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [55]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

# use filter/lambda funciton

result = list(filter(lambda name : name.strip() if name != ' ' else False, places))
print(result)


['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [88]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

# lambda function
# use key  for .sort function

author.sort(key=lambda x : x.lower().split()[-1]) # lower has to go before split
print(author)


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [180]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]



# use map and lambda
placesInF = list(map(lambda x : (x[0], (9/5) * x[1] + 32), places)) # can create tuple with lambda by putting (num,num) in func

print(placesInF)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [2]:
# 1 , 1 = 2, then 1 , 2 = 3, 2, 3 = 5, 3,5 = 8 5,8 = 13

def fib(num): # Regular Fib function
    if num == 0:
        return 1
    elif num == 1:
        return 1
    elif num == 2:
        return 2
    else:
        return fib(num-1) + fib(num-2)

def iteration(n): # Looping through iterations 
    for x in range(n+1): # Start by creating range through the number we want (5) 
        print('Iteration({x}): {r}'.format(x=x,r=fib(x))) #.format used as f-string (but can not use f, and have to define ALL variables), but can put in own functions/variables 






print(iteration(5))

Iteration(0): 1
Iteration(1): 1
Iteration(2): 2
Iteration(3): 3
Iteration(4): 5
Iteration(5): 8
None
